File: web/dart/blocks/expression.dart
/*
 * NetTango
 * Copyright (c) 2017 Michael S. Horn, Uri Wilensky, and Corey Brady
 *
 * Northwestern University
 * 2120 Campus Drive
 * Evanston, IL 60613
 * http://tidal.northwestern.edu
 * http://ccl.northwestern.edu

 * This project was funded in part by the National Science Foundation.
 * Any opinions, findings and conclusions or recommendations expressed in this
 * material are those of the author(s) and do not necessarily reflect the views
 * of the National Science Foundation (NSF).
 */
part of NetTango;


/// TODO
///   limit length of inline expressions on blocks.
class Expression {

  ExpressionBuilder builder;

  String name = null;

  String type = "num";

  String format = null;

  bool get isRoot => (builder.root == this);

  bool get isEmpty => (name == null);

  bool get isUnary => (children.length == 1);

  bool get isBinary => (children.length == 2);

  bool get hasChildren => (children.isNotEmpty);

  bool get isChildless => (children.length == 0);

  List<Expression> children = new List<Expression>();


  Expression(this.builder, this.type);


  void displayString(StringBuffer out) {
    if (isUnary) {
      if (!isRoot) out.write("(");
      out.write("$name ");
      children[0].displayString(out);
      if (!isRoot) out.write(")");
    }
    else if (isBinary) {
      if (!isRoot) out.write("(");
      children[0].displayString(out);
      out.write(" $name ");
      children[1].displayString(out);
      if (!isRoot) out.write(")");
    }
    else if (name != null) {
      out.write("$name");
    }
  }


  Map toJSON() {
    Map data = { "name" : name, "type" : type };
    if (hasChildren) {
      data["children"] = [];
      for (Expression child in children) {
        data["children"].add(child.toJSON());
      }
    }
    if (format != null) data["format"] = format;
    return data;
  }


  void fromJSON(Map json) {
    name = toStr(json['name']);
    type = toStr(json['type'], "num");
    if (json['format'] != null) {
      format = json['format'];
    }
    children.clear();
    if (json['children'] is List) {
      for (var c in json['children']) {
        Expression child = new Expression(builder, c['type']);
        children.add(child);
        child.fromJSON(c);
      }
    }
  }


  // test to see if the current children match arg list?
  // if so, leave them alone rather than replace them
  bool childMismatch(List args) {
    if (args == null) return children.isNotEmpty;
    if (children.length != args.length) return true;
    for (int i=0; i<args.length; i++) {
      if (args[i] != children[i].type) return true;
    }
    return false;
  }


  void setChildren(List args) {
    bool childless = isChildless;

    if (childMismatch(args)) {
      children.clear();
      if (args != null) {
        for (int i=0; i<args.length; i++) {

          // chain first expression?
          if (i == 0 && childless && args[i] == type) {
            children.add(new Expression(builder, args[i]) .. name = name);
          } else {
            children.add(new Expression(builder, args[i]));
          }
        }
      }
    }
  }


  void appendOperator(DivElement parent) {
    DivElement div = new DivElement()
      .. innerHtml = "$name"
      .. classes.add("nt-expression-text")
      .. classes.add("editable")
      .. classes.add("$type");

    div.onClick.listen((e) {
      openPulldown(div);
      e.stopPropagation();
    });
    electricBrace(div, parent);
    parent.append(div);
  }


  void electricBrace(DivElement curr, DivElement parent) {
    curr.onMouseEnter.listen((e) { parent.classes.add("highlight"); });
    curr.onMouseLeave.listen((e) { parent.classes.remove("highlight"); });
  }


  void appendParen(DivElement parent, bool left) {
    DivElement paren = new DivElement()
      .. innerHtml = left ? "(" : ")"
      .. classes.add("nt-expression-text")
      .. classes.add("parenthesis");
    electricBrace(paren, parent);
    parent.append(paren);
  }


  void appendNumber(DivElement parent) {
    name = toNum(name, 0).toString();
    NumberInputElement input = new NumberInputElement()
      .. className = "nt-number-input"
      .. value = name
      .. step = "1";
    input.onChange.listen((e) {
      name = input.value;
      if (name == "") {
        name = "0";
        input.value = "0";
      }
    });
    parent.append(input);
  }


  bool get isNum {
    if (name != null) {
      return num.parse(name, (e) => null) != null;
    }
    return false;
  }


  void renderHtml(DivElement parent) {
    DivElement div = new DivElement() .. className = "nt-expression";
    if ((isNum || isEmpty) && type == "num") {
      appendNumber(div);
    }
    else if (isEmpty) {
      div.classes.add("empty");
      div.appendHtml("<small>&#9660;</small>");
    }
    else if (isUnary) {
      appendParen(div, true);
      appendOperator(div);
      children[0].renderHtml(div);
      appendParen(div, false);
    }
    else if (isBinary) {
      appendParen(div, true);
      children[0].renderHtml(div);
      appendOperator(div);
      children[1].renderHtml(div);
      appendParen(div, false);
    }
    else {
      div.appendHtml("<div class='nt-expression-text $type'>$name</div>");
    }
    if (isChildless) {
      div.classes.add("editable");
      div.onClick.listen((e) {
        openPulldown(div);
        e.stopPropagation();
      });
    }
    parent.append(div);
  }


 //-------------------------------------------------------------
 // Creates an expression pulldown menu
 //-------------------------------------------------------------
  void openPulldown(Element expander) {
    querySelectorAll('.nt-pulldown-menu').forEach((el) => el.remove());
    DivElement hmenu = new DivElement() .. classes.add('nt-pulldown-menu');

    // ---------------  expressions ---------------------
    _addMenuItems(hmenu, builder.expressions);

    // ---------------  variables ---------------------
    if (builder.variables.isNotEmpty) hmenu.appendHtml("<hr>");
    _addMenuItems(hmenu, builder.variables);

    // ---------------  clear button ---------------------
    hmenu.appendHtml("<hr>");
    AnchorElement link = new AnchorElement(href : "#")
      .. innerHtml = "Clear"
      .. className = "clear";
    hmenu.append(link);
    link.onClick.listen((e) {
      hmenu.remove();
      name = null;
      children.clear();
      format = null;
      builder.renderHtml();
      e.stopPropagation();
      e.preventDefault();
    });

    expander.append(hmenu);
  }


  void _addMenuItems(DivElement hmenu, List items) {
    for (var item in items) {
      if (item['type'] == type) {
        AnchorElement link = new AnchorElement(href : "#")
          .. innerHtml = "${item['name']}";
        hmenu.append(link);
        link.onClick.listen((e) {
          hmenu.remove();
          setChildren(item['arguments']);
          name = item['name'];
          type = item['type'];
          format = item['format'];
          builder.renderHtml();
          e.stopPropagation();
          e.preventDefault();
        });
      }
    }
  }

}


class ExpressionBuilder {

  CodeWorkspace workspace;

  Element parent;

  Expression root;

  List get expressions => workspace.expressions;

  List get variables => workspace.variables;


  ExpressionBuilder(this.workspace, String type) {
    root = new Expression(this, type);
  }


  String toString() {
    StringBuffer out = new StringBuffer();
    root.displayString(out);
    return out.toString();
  }


  Map toJSON() {
    return root.toJSON();
  }


  void fromJSON(var json) {
    if (json is Map) {
      root.fromJSON(json);
    } else if (json != null) {
      root.name = json.toString();
    }
  }


  void open(String parentSelector) {
    this.parent = querySelector(parentSelector);
    renderHtml();
  }


  renderHtml() {
    if (parent != null && root != null) {
      parent.children.clear();
      root.renderHtml(parent);
    }
  }
}

File: src/blocks/expressions/expression.ts
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

import { ExpressionDefinition, ExpressionInput } from "../../types/types"
import { NumUtils } from "../../utils/num-utils"
import { StringBuffer } from "../../utils/string-buffer"
import { StringUtils } from "../../utils/string-utils"
import { ExpressionBuilder } from "./expression-builder"

/// TODO
///   limit length of inline expressions on blocks.
class Expression {

  e: ExpressionInput
  builder: ExpressionBuilder
  children: Expression[]

  get isRoot(): boolean { return this.builder.root === this }
  get isEmpty(): boolean { return this.e.name === null }
  get isUnary(): boolean { return this.e.children.length === 1 }
  get isBinary(): boolean { return this.e.children.length === 2 }
  get hasChildren(): boolean { return this.e.children.length > 0 }
  get isChildless(): boolean { return this.e.children.length === 0 }
  get isNum(): boolean {
    if (this.e.name !== null) {
      return !Number.isNaN(Number.parseFloat(this.e.name))
    }
    return false
  }

  constructor(builder: ExpressionBuilder, type: 'num' | 'bool') {
    this.builder = builder
    this.e = Expression.createEmptyExpression(type)
    this.children = []
  }

  static getDefaultValue(type: 'num' | 'bool'): '0' | 'false' {
    return (type === 'num') ? '0' : 'false'
  }

  static createEmptyExpression(type: 'num' | 'bool'): ExpressionInput {
    return {
      name: Expression.getDefaultValue(type)
    , type: type
    , format: null
    , children: []
    }
  }

  static cloneExpressionInput(source: ExpressionInput): ExpressionInput {
    const clone = Expression.createEmptyExpression(source.type)
    source.children.forEach( (child) => clone.children.push(Expression.cloneExpressionInput(child) ))
    return clone
  }

  static clone(builder: ExpressionBuilder, source: Expression): Expression {
    const c = new Expression(builder, source.e.type)
    c.e = Expression.cloneExpressionInput(source.e)
    source.children.forEach( (child) => c.children.push(Expression.clone(c.builder, child)) )
    return c
  }

  displayString(out: StringBuffer): void {
    if (this.isUnary) {
      if (!this.isRoot) {
        out.write("(")
      }
      out.write(`${this.e.name} `)
      this.children[0].displayString(out)
      if (!this.isRoot) {
        out.write(")")
      }
    }
    else if (this.isBinary) {
      if (!this.isRoot) {
        out.write("(")
      }
      this.children[0].displayString(out)
      out.write(` ${this.e.name} `)
      this.children[1].displayString(out)
      if (!this.isRoot) {
        out.write(")")
      }
    }
    else if (this.e.name !== null) {
      out.write(`${this.e.name}`)
    }
  }

  // test to see if the current children match arg list?
  // if so, leave them alone rather than replace them
  childMismatch(args: ("num" | "bool")[]): boolean {
    if (args === null) {
      return this.children.length !== 0
    }
    if (this.children.length !== args.length) {
      return true
    }
    for (var i = 0; i < args.length; i++) {
      if (args[i] !== this.children[i].e.type) {
        return true
      }
    }
    return false
  }

  setChildren(args: ("num" | "bool")[]): void {
    const childless = this.isChildless

    if (this.childMismatch(args)) {
      this.children.splice(0)
      this.e.children.splice(0)
      if (args !== null) {
        for (var i = 0; i < args.length; i++) {
          // chain first expression?
          const e = new Expression(this.builder, args[i])
          if (i === 0 && childless && args[i] === this.e.type) {
            e.e.name = this.e.name
            this.children.push(e)
            this.e.children.push(e.e)
          } else {
            this.children.push(e)
            this.e.children.push(e.e)
          }
        }
      }
    }
  }

  appendOperator(parent: HTMLDivElement): void {
    const div = document.createElement("div")
    div.innerHTML = StringUtils.toStr(this.e.name, "")
    div.classList.add("nt-expression-text")
    div.classList.add("editable")
    div.classList.add(this.e.type)

    div.addEventListener("click", (e) => {
      this.openPulldown(div)
      e.stopPropagation()
    })
    this.electricBrace(div, parent)
    parent.append(div)
  }

  electricBrace(curr: HTMLDivElement, parent: HTMLDivElement): void {
    curr.addEventListener("mouseenter", (e) => { parent.classList.add("highlight") })
    curr.addEventListener("mouseleave", (e) => { parent.classList.remove("highlight") })
  }

  appendParen(parent: HTMLDivElement, left: boolean): void {
    const paren = document.createElement("div")
    paren.innerHTML = left ? "(" : ")"
    paren.classList.add("nt-expression-text")
    paren.classList.add("parenthesis")
    this.electricBrace(paren, parent)
    parent.append(paren)
  }

  appendNumber(parent: HTMLDivElement): void {
    this.e.name = NumUtils.toNum(this.e.name, 0).toString()
    const input = document.createElement("input")
    input.type = "number"
    input.className = "nt-number-input"
    input.value = this.e.name
    input.step = "1"
    input.addEventListener("change", (e) => {
      this.e.name = input.value
      if (this.e.name === "") {
        this.e.name = "0"
        input.value = "0"
      }
    })
    parent.append(input)
  }

  renderHtml(parent: Element): void {
    const div = document.createElement("div")
    div.className = "nt-expression"
    if ((this.isNum || this.isEmpty) && this.e.type === "num") {
      this.appendNumber(div)
    }
    else if (this.isEmpty) {
      div.classList.add("empty")
      div.insertAdjacentHTML("beforeend", "<small>&#9660;</small>")
    }
    else if (this.isUnary) {
      this.appendParen(div, true)
      this.appendOperator(div)
      this.children[0].renderHtml(div)
      this.appendParen(div, false)
    }
    else if (this.isBinary) {
      this.appendParen(div, true)
      this.children[0].renderHtml(div)
      this.appendOperator(div)
      this.children[1].renderHtml(div)
      this.appendParen(div, false)
    }
    else {
      div.insertAdjacentHTML("beforeend", `<div class='nt-expression-text ${this.e.type}'>${this.e.name}</div>`)
    }
    if (this.isChildless) {
      div.classList.add("editable")
      div.addEventListener("click", (e) => {
        this.openPulldown(div)
        e.stopPropagation()
      })
    }
    parent.append(div)
  }

  //-------------------------------------------------------------
  // Creates an expression pulldown menu
  //-------------------------------------------------------------
  openPulldown(expander: Element): void {
    document.querySelectorAll('.nt-pulldown-menu').forEach( (el) => el.remove() )
    const hmenu = document.createElement("div")
    hmenu.classList.add('nt-pulldown-menu')

    // ---------------  expressions ---------------------
    this._addMenuItems(hmenu, this.builder.expressions)

    // ---------------  variables ---------------------
    if (this.builder.variables.length > 0) hmenu.insertAdjacentHTML("beforeend", "<hr>")
    // TODO: get this working again once variables are implemented
    // _addMenuItems(hmenu, builder.variables)

    // ---------------  clear button ---------------------
    hmenu.insertAdjacentHTML("beforeend", "<hr>")
    const link = document.createElement("a")
    link.href = "#"
    link.innerHTML = "Clear"
    link.className = "clear"
    hmenu.append(link)
    link.addEventListener("click", (e) => {
      hmenu.remove()
      this.e.name = Expression.getDefaultValue(this.e.type)
      this.children.splice(0)
      this.e.children.splice(0)
      this.e.format = null
      this.builder.renderHtml()
      e.stopPropagation()
      e.preventDefault()
    })

    expander.append(hmenu)
  }

  _addMenuItems(hmenu: HTMLDivElement, items: ExpressionDefinition[]): void {
    for (const item of items) {
      if (item.type === this.e.type) {
        const link = document.createElement("a")
        link.href = "#"
        link.innerHTML = `${item.name}`
        hmenu.append(link)
        link.addEventListener("click", (e) => {
          hmenu.remove()
          this.setChildren(item.arguments)
          this.e.name = item.name
          this.e.type = item.type
          this.e.format = item.format
          this.builder.renderHtml()
          e.stopPropagation()
          e.preventDefault()
        })
      }
    }
  }

}

export { Expression }
